### PR TITLE
Connect command.py with common.py and create new methods

### PR DESCRIPTION
**Pull request by giannissc**

### CODE REVIEW

**Comment by giannissc** [comment]
I can confirm that the mark ready for review button is not available when you first create the PR so I am pinging you here @voneiden

**Comment by giannissc** [comment]
Remaining tasks for this are to:
- Create documentation for command.py
- Create tests for new methods

Should I include those on this PR or should they be on a separate one? @voneiden

EDIT: Converting PR to draft until resolved

**Comment by voneiden** [comment]
Yes, please include tests in the PR, currently I've set sonarcloud to require a 65% coverage before merge is allowed. Including documentation here or in a separate PR is fine either way. 

I've tried adding a CODEOWNERS file in b3e5e1f - if that works it should assign me as a reviewer automatically when a PR is opened (excluding drafts).

**Comment by giannissc** [comment]
I am thinking of changing the structure of Command slightly. One thing that I don't understand is why does to_gcode() return (str, Vector)?

The second is why must Command have an attribute `end`

**Comment by voneiden** [comment]
> why does to_gcode() return (str, Vector)?

That might just be legacy baggage. See if you can get rid of it. Technically you should be able to use the end position of the previous command but..

> The second is why must Command have an attribute end

Probably because previously all (?) commands have been motion commands. Might make sense to create a MotionCommand(Command) subclass that defines the end vector. However now if you make it possible to have commands without an end vector, you'll need to consider the first case. Possibly keep track of the last known end position here:

https://github.com/voneiden/cq-cam/blob/b3e5e1f3942aeadd211387f2829129367b4e9767/src/cq_cam/fluent.py#L39

The previous position is used to optimize the gcode.

```
G0 X1 Y1 Z1
G0 X1 Y2 Z1
```

the coordinates (and G value) that did not change and are not relevant can be skipped producing

```
G0 X1 Y1 Z1
Y2
```

While technically it is not necessary to do this, it can be a significant file size reduction. 

ReferencePosition class seems to be obsolete, feel free to delete.

**Comment by giannissc** [comment]
Ok now I understand the reasoning! I have played a bit with removing but it makes thing unnecessarily complicated. I was also thinking of creating to abstract sub-classes from Command a MotionCommand (as you suggested) and a ConfigCommand

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py]: Good catch with a new catch:

I learned this quite recently myself too, but `Tuple[...]` is deprecated ( https://docs.python.org/3/library/typing.html#typing.Tuple ) since Python 3.9. The new official way to do it now is  `tuple[...]`  ([PEP 585](https://peps.python.org/pep-0585/))

Comment (giannissc) [reply src/cq_cam/command.py]: Good to know I will do this change as well!

**Comment by giannissc** [comment]
Just added the test cases. I need to add the documentation for the module and I will mark it as ready for review

**Comment by giannissc** [comment]
Why is sonar cloud failing every time? Is there anything that I can do on my end to fix it?

**Comment by voneiden** [comment]
> Why is sonar cloud failing every time? Is there anything that I can do on my end to fix it?

This is a limitation of SonarCloud and forks on Github ( https://portal.productboard.com/sonarsource/1-sonarcloud/c/50-sonarcloud-analyzes-external-pull-request ). Basically the PR checks are run in the context of your fork repository, so it does not have access to the required SONAR_TOKEN secret key that is configured in my repository.

If you're interested, you could try creating an account at SonarCloud and I'll try adding you to my personal organization ( https://sonarcloud.io/organizations/voneiden/ ). After you've created an account

1) Head over to https://sonarcloud.io/account/security  and create a new token
2) Head over to https://github.com/giannissc/cq-cam/settings/secrets/actions and add the token as a repository secret with the name SONAR_TOKEN

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py]: Python has quite a few ways to build strings..  `+=`  is fine from readability perspective, but this line is so long it could benefit from some kind of alternative approach. When concatenating a lot of strings, `+` is the slowest (doesn't matter in this case). 

One option is [literal string concatenation](https://docs.python.org/3/reference/lexical_analysis.html#string-literal-concatenation) and it could look like this with f-strings:

```python
return (
        f"{DistanceMode.ABSOLUTE.to_gcode()} "
        f"{WorkOffset.OFFSET_1.to_gcode()} "
        f"{PlannerControlMode.BLEND.to_gcode()} "
        f"{SpindleControlMode.MAX_SPINDLE_SPEED.to_gcode()} "
        f"{WorkPlane.XY.to_gcode()} "
        f"{FeedRateControlMode.UNITS_PER_MINUTE.to_gcode()}\n"
        f"{LengthCompensation.OFF.to_gcode()} "
        f"{RadiusCompensation.OFF.to_gcode()} "
        f"{CannedCycle.CANCEL.to_gcode()}\n"
        f"{Unit.METRIC.to_gcode()}\n"
        f"{HomePosition.HOME_2.to_gcode()}\n"
    )
```

Downside for readability is that it is slightly hard to see at a glance where each line ends. Linefeeds could also be on separate lines

```python
return (
        f"{DistanceMode.ABSOLUTE.to_gcode()} "
        f"{WorkOffset.OFFSET_1.to_gcode()} "
        f"{PlannerControlMode.BLEND.to_gcode()} "
        f"{SpindleControlMode.MAX_SPINDLE_SPEED.to_gcode()} "
        f"{WorkPlane.XY.to_gcode()} "
        f"{FeedRateControlMode.UNITS_PER_MINUTE.to_gcode()}"
        f"\n"
        f"{LengthCompensation.OFF.to_gcode()} "
        f"{RadiusCompensation.OFF.to_gcode()} "
        f"{CannedCycle.CANCEL.to_gcode()}"
        f"\n"
        f"{Unit.METRIC.to_gcode()}"
        f"\n"
        f"{HomePosition.HOME_2.to_gcode()}"
        f"\n"
    )
```

Another option is to go more in the direction of functional programming and use iterable joins. I'm a bit of a sucker for this approach.

```python

return "\n".join(
    (
        " ".join(
            (
                DistanceMode.ABSOLUTE.to_gcode(),
                WorkOffset.OFFSET_1.to_gcode(),
                PlannerControlMode.BLEND.to_gcode(),
                SpindleControlMode.MAX_SPINDLE_SPEED.to_gcode(),
                WorkPlane.XY.to_gcode(),
                FeedRateControlMode.UNITS_PER_MINUTE.to_gcode(),
            )
        ),
        " ".join(
            (
                LengthCompensation.OFF.to_gcode(),
                RadiusCompensation.OFF.to_gcode(),
                CannedCycle.CANCEL.to_gcode(),
            )
        ),
        Unit.METRIC.to_gcode(),
        HomePosition.HOME_2.to_gcode(),
    )
```

Side note, Python 3.11 introduces [StrEnum](https://docs.python.org/3/library/enum.html#enum.StrEnum), could be nice to use to get rid of to_gcode() calls in the future.

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py]: PEP-8 @ https://peps.python.org/pep-0008/#programming-recommendations

> Comparisons to singletons like None should always be done with `is` or `is not`, never the equality operators.

```suggestion
        if self.coolant is not None:
```

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py:181]: Few thoughts.. 

1) Ideally base class should only implement stuff that is common to all subclasses. SafetyBlock for example doesn't use any of these three attributes defined here.

2) `spindle` and `coolant` here are slightly suspicious, BUT I'm not entirely sure how you intend to integrate this in practice. The main concern is where is the "master data". For example if the user has enabled coolant for an operation and that state is stored in the Operation class, is it redundant to store it also in the ConfigCommand? Rhetorical question. :smile:

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py]: Accidental tuple? `(None,)`

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py]:
```suggestion
        if self.spindle is not None:
```

File: src/cq_cam/command.py
Comment (voneiden) [inline src/cq_cam/command.py]:
```suggestion
        if self.coolant is not None:
```

**Review by voneiden — CHANGES_REQUESTED** [review]

File: src/cq_cam/command.py
Comment (giannissc) [reply src/cq_cam/command.py:181]: Spindle here needs to be unique between commands as it can be specific to an operation. Coolant's master data always lives in job and shouldn't change between operations

File: src/cq_cam/command.py
Comment (giannissc) [reply src/cq_cam/command.py]: I will use StrEnum is another PR I think. I am currently on Python 3.10

File: src/cq_cam/command.py
Comment (giannissc) [reply src/cq_cam/command.py]: Yeah!

File: src/cq_cam/command.py
Comment (giannissc) [reply src/cq_cam/command.py:181]: I just didn't implement the interface of config command to have a reference to Job as I didn't want a different interface to MotionCommand.

**Comment by giannissc** [comment]
I will do sonar cloud soon to get rid of all the errors

File: src/cq_cam/command.py
Comment (voneiden) [reply src/cq_cam/command.py]: Yeah just pointing it out. I'm pretty sure we need to support py3.10 for a while into the future so can't use StrEnum yet anyway.

**Comment by giannissc** [comment]
SONAR_TOKEN is set up. What do you need to add me to your organisation? @voneiden

**Comment by voneiden** [comment]
@giannissc bleh, didn't work, looks like the PR workflow has no access to secrets from your account either.   I need to see if I can still try working around this somehow.

But I've added you to the sonarcloud org anyway.

**Comment by giannissc** [comment]
Maybe I need to make another commit to retrigger the CI pipeline?

**Comment by voneiden** [comment]
Nah it's not going to work. I think I will try using the automatic analysis feature of sonarcloud, it shouldn't have this restriction although code coverage will be lost (but suppose that's not a big deal since there's codecov)

File: src/cq_cam/command.py
Comment (voneiden) [reply src/cq_cam/command.py:181]: OK, I see it now. I've left further comments in fluent.py

File: src/cq_cam/fluent.py
Comment (voneiden) [inline src/cq_cam/fluent.py:41]: Technically this works just fine, but initializing instance attributes outside of `__init__` sets my spider senses tingling. 

It is fairly harmless in this particular small scope, but risks becoming unsustainable if this approach is used liberally throughout the code base. Slippery spaghetti slope :warning: 


I'll think about this still a bit, but I think the appropriate approach might be to make the `to_gcode` more flexible with `**kwargs`. This also applies to the `ConfigCommand` discussion above. Something like this maybe

```python
class BaseCommand(ABC):
    @abstractmethod
    def to_gcode(self, x=None, y=None, z=None):
        pass


class SpecialCommand(BaseCommand):
    def to_gcode(self, x=None, **kwargs):
        # Do something with x
        pass
```

Comment (voneiden) [reply src/cq_cam/fluent.py:41]: With the above approach this line could be something like
```suggestion
            gcode, position = command.to_gcode(operation=self, previous_command=previous_command, start=position)
```

This way it'd be possible to access both the operation and the job (`operation.job`)  in any `to_gcode`  without needing to store copies of top-level state inside Commands.

**Review by voneiden — CHANGES_REQUESTED** [review]
Alrighty, a little bit of architectural tweaking requested still. Otherwise everything's looking good.

Once that's settled and the code is ready to be merged I'd still like to clean up the commit history a bit. Are you interested giving it a shot or would you prefer me to handle it? I'm not sure what your experience level is with git (for example rebasing), but I can certainly try giving you some pointers to get going.

**Comment by giannissc** [comment]
> Alrighty, a little bit of architectural tweaking requested still. Otherwise everything's looking good.
> 
> Once that's settled and the code is ready to be merged I'd still like to clean up the commit history a bit. Are you interested giving it a shot or would you prefer me to handle it? I'm not sure what your experience level is with git (for example rebasing), but I can certainly try giving you some pointers to get going.

I am happy to do the rebase and I would appreciate the pointers. I make the changes today

**Comment by voneiden** [comment]
You use vscode, right? Maybe this is a good primer https://blog.delpuppo.net/why-i-love-gitlens-in-my-vscode-interactive-rebase

I suggest you create a branch and try out rebasing there. That way it's easier to just throw away the branch if things go south. I have a local copy of this PR too anyway, so no worries. 

What I'd like to see is that the commit history is tidied up a bit

1) Squash/fixup strongly related commits together unless keeping them separate makes sense. This means minor fix commits get squashed with their original feature commit.
2) No reverse merge commits (like [Merge branch 'voneiden:unstable' into unstable](https://github.com/voneiden/cq-cam/pull/28/commits/99f517391e6a895b7088552b797663f9e896559d)) unless absolutely necessary. 

To avoid 2 in the first place, during development you can use `git rebase` instead of `git merge` and/or `git pull --rebase` instead of `git pull`.

Feel free to ask and feel free to throw in the towel. I know git can be overwhelming.

**Comment by giannissc** [comment]
I am thinking that I won't be using kwargs for the gcode methods as I will be removing them on the next PR. I will write more of my thoughts into the discussion @voneiden

**Pull request closed without merging**
